hwt: spe: initial MODE_CPU support

zdleaf opened the pull request:
n/a

zdleaf committed:
hwt: spe: initial MODE_CPU support

zdleaf committed:
hwt: disable config ioctl once started

Once running we shouldn't be able to use the config ioctl

zdleaf committed:
hwt: refactor IOC_BUFPTR_GET

Make this a bit less backend specific by renaming curpage - SPE is not
working with pages or page offsets.

Plus SPE need an extra *data field to pass various flags.

zdleaf committed:
hwt: usr: improved error msgs for bufsize

bukinr commented on sys/arm64/arm64/locore.S:
```diff
@@ -397,6 +397,12 @@ LENTRY(enter_kernel_el)
+	/* Set the Monitor Debug Configuration Register */
+	mrs	x2, mdcr_el2
+	/* Enable SPE at EL1 */
+	orr	x2, x2, (3 << 12)
```
is this bit available on all arm64 machines? (or should we enable SPE at EL1 by request?)
also we need to unmagic the bits

zxombie replied:
This should be safe as on earlier revisions of the Architecture those bits did nothing.

zdleaf replied:
Fixed. This value will need to change once Andy's VHE patches are in or close to in.

zxombie commented on sys/dev/hwt/hwt_hook.c:
```diff
@@ -282,8 +284,17 @@ hwt_hook_handler(struct thread *td, int func, void *arg)
+#ifdef __aarch64__
+		if (hwt_log_pid) {
+			if (func == HWT_SWITCH_IN)
+				WRITE_SPECIALREG(CONTEXTIDR_EL1_REG, p->p_pid);
```
We should update `contextidr_el1` in the thread switcher. I didn't do it previously as I was unsure what the register was for.

zdleaf replied:
Okay. Maybe we should introduce some kind of additional config option to enable/disable this, although I don't know if it's useful outside of hardware tracing. Otherwise we can enable/disable based on if HWT_HOOKS since it adds additional instructions to context switching.

zxombie replied:
There is also a use by the debug logic, e.g. breakpoints & watchpoints, although I'm not sure how useful it is there.

zdleaf replied:
I will add a separate conf option to toggle it then.

zdleaf replied:
I've added a kernel conf option + sysctl switch for that. See [arm64: write PID in CONTEXTIDR_EL1 on ctx switch](https://github.com/CTSRD-CHERI/freebsd-morello/pull/10/commits/1fffa1f5c4c56963e526401a30257768e01e879e). I'll send this for a separate review on Phab.

zdleaf committed:
hwt: spe: move to sys/arm64/spe

zdleaf committed:
hwt: spe: kernel module

zdleaf committed:
arm64: add MDCR_EL2 reg

Monitor Debug Configuration Register provides EL2 configuration options
for self-hosted debug and the Performance Monitors Extension.

Sponsored by:	Arm Ltd

zdleaf committed:
hwt: spe: unmagic MDCR bits

zdleaf committed:
hwt: spe: refactor CONTEXTIDR_EL1 writes

zdleaf committed:
hwt: spe: sort includes

zdleaf commented on sys/arm64/arm64/sys_machdep.c:
```diff
@@ -80,3 +83,10 @@ sysarch(struct thread *td, struct sysarch_args *uap)
+
+#ifdef PID_IN_CONTEXTIDR
+bool arm64_pid_in_contextidr = false;
+SYSCTL_BOOL(_machdep, OID_AUTO, pid_in_contextidr, CTLFLAG_RW,
```
I'm not sure machdep is the correct place for this. The debug sysctl was another place I considered.

bukinr commented on sys/dev/hwt/hwt_hook.c:
```diff
@@ -230,13 +229,11 @@ hwt_hook_thread_create(struct thread *td)
 
 	/* Step 2. Allocate some memory without holding ctx ref. */
-	error = hwt_thread_alloc(&thr, path, bufsize, kva_req);
+	error = hwt_thread_alloc(&thr, path, ctx);
```
per discussion in IRC, could we restore this line? dereferencing ctx may not work as we released ref a line before

zdleaf replied:
Done

bukinr reviewed (changes requested):
looks good!

zdleaf committed:
arm64: write PID in CONTEXTIDR_EL1 on ctx switch

Debug and trace features such as Statistical Profiling Extension (SPE)
use the CONTEXTIDR_EL1 register to get the PID of the current process.

Add a sysctl switch to toggle writing the current PID into this register
in the thread switcher.

To make use of the feature, the following sysctl switch must be set:

    sysctl machdep.pid_in_contextidr=1

Kernel code can also toggle the sysctl by writing directly to the global
var behind the sysctl arm64_pid_in_contextidr:

    extern bool arm64_pid_in_contextidr;

Sponsored by:	Arm Ltd

zdleaf committed:
hwt: fix kva_req

Setting ctx->kva_req in backend_init() is too late - we've already
allocated buffers by that point

Save kva_req in struct hwt_backend instead so it's available early

zdleaf committed:
hwt: spe: define spe specific taskqueue

Coresight doesn't require a taskqueue and Intel PT currently defines
it's own anyway. Create a taskqueue for SPE to use and remove out of
date comments about it.

zdleaf commented:
- Rebase
- Removed kernel conf option for CONTEXTIDR_EL1
- Sent `arm64:` patches upstream:
    - https://reviews.freebsd.org/D46191
    - https://reviews.freebsd.org/D46192
- Fixed kva_req/bufsize arg

zdleaf commented:
- Fix comment in [hwt: spe: unmagic MDCR bits](https://github.com/CTSRD-CHERI/freebsd-morello/pull/10/commits/cec0b86f61cbc0c735d63920b3809e959e083d87)
